File: _/Source code/CoherentBank/net/src/BranchTerminal.WPF/Domain/Transaction.cs
using System;

using Seovic.Core;

using Tangosol.IO.Pof;
using Tangosol.Net.Cache;

namespace Seovic.Samples.Bank.Domain
{
    public class Transaction : IEntity<Transaction.TransactionId>, IPortableObject
    {
        #region Constructors and factory methods

        public Transaction()
        {}

        private Transaction(TransactionId id, TransactionType type,
                            DateTime time, String description, 
                            Money amount, Money balance)
        {
            this.id          = id;
            this.type        = type;
            this.time        = time;
            this.description = description;
            this.amount      = amount;
            this.balance     = balance;
        }

        public static Transaction Create(long accountId, long transactionNumber,
                                         TransactionType type, String description, 
                                         Money amount, Money balance)
        {
            return new Transaction(new TransactionId(accountId, transactionNumber), type,
                                   DateTime.Now, description, amount, balance);
        }

        #endregion

        #region Properties

        public TransactionId Id
        {
            get { return id; }
        }

        public TransactionType Type
        {
            get { return type; }
        }

        public DateTime Time
        {
            get { return time; }
        }

        public string Description
        {
            get { return description; }
        }

        public Money Amount
        {
            get { return amount; }
        }

        public Money Balance
        {
            get { return balance; }
        }

        public Money DepositAmount
        {
            get
            {
                return TransactionType.DEPOSIT.Equals(Type) 
                    ? Amount
                    : null;
            }
        }

        public Money WithdrawalAmount
        {
            get
            {
                return TransactionType.WITHDRAWAL.Equals(Type) 
                    ? Amount
                    : null;
            }
        }

        #endregion

        #region Object methods

        /// <summary>
        /// Returns a string representation of the current object.
        /// </summary>
        /// <returns>
        /// String that represents the current object.
        /// </returns>
        public override string ToString()
        {
            return string.Format("Id: {0}, Type: {1}, Time: {2}, Description: {3}," + 
                " Amount: {4}, Balance: {5}", id, type, time, description, amount, balance);
        }

        #endregion

        #region IPortableObject implementation

        public void WriteExternal(IPofWriter writer)
        {
            writer.WriteObject(0, id);
            writer.WriteObject(1, type);
            writer.WriteDate  (2, time);
            writer.WriteString(3, description);
            writer.WriteObject(4, amount);
            writer.WriteObject(5, balance);
        }

        public void ReadExternal(IPofReader reader)
        {
            id          = (TransactionId) reader.ReadObject(0);
            type        = (TransactionType) reader.ReadObject(1);
            time        = reader.ReadDate(2);
            description = reader.ReadString(3);
            amount      = (Money) reader.ReadObject(4);
            balance     = (Money) reader.ReadObject(5);
        }


        #endregion

        #region Inner class: TransactionId

        public class TransactionId : IKeyAssociation, IPortableObject
        {
            #region Constructors

            public TransactionId()
            {
            }

            public TransactionId(long accountId, long transactionNumber)
            {
                this.accountId         = accountId;
                this.transactionNumber = transactionNumber;
            }

            #endregion

            #region Properties

            public long AccountId
            {
                get { return accountId; }
            }

            public long TransactionNumber
            {
                get { return transactionNumber; }
            }

            #endregion

            #region IKeyAssociation implementation

            public object AssociatedKey
            {
                get { return accountId; }
            }

            #endregion

            #region IPortableObject implementation

            public void ReadExternal(IPofReader reader)
            {
                accountId         = reader.ReadInt64(0);
                transactionNumber = reader.ReadInt64(1);
            }

            public void WriteExternal(IPofWriter writer)
            {
                writer.WriteInt64(0, accountId);
                writer.WriteInt64(1, transactionNumber);
            }

            #endregion

            #region Data members

            private long accountId;
            private long transactionNumber;

            #endregion
        }

        #endregion

        #region Data members

        private TransactionId id;
        private TransactionType type;
        private DateTime time;
        private String description;
        private Money amount;
        private Money balance;

        #endregion
    }
}

File: _/Source code/CoherentBank/java/src/com/seovic/samples/bank/domain/Transaction.java
package com.seovic.samples.bank.domain;


import com.seovic.coherence.io.pof.AbstractPofSerializer;
import com.seovic.core.Entity;

import com.tangosol.io.pof.PortableObject;
import com.tangosol.io.pof.PofReader;
import com.tangosol.io.pof.PofWriter;
import com.tangosol.net.cache.KeyAssociation;

import java.util.Date;
import java.io.IOException;
import org.codehaus.jackson.annotate.JsonIgnore;


/**
 * @author Aleksandar Seovic  2009.12.01
 */
public class Transaction
        implements Entity<Transaction.Id>
    {
    // ---- constructors and factory methods --------------------------------

    public Transaction(Id id, TransactionType type,
                       Date time, String description, Money amount,
                       Money balance)
        {
        m_id          = id;
        m_type        = type;
        m_time        = time;
        m_description = description;
        m_amount      = amount;
        m_balance     = balance;
        }

    static Transaction create(long accountId, long transactionNumber,
                              TransactionType type, String description,
                              Money amount, Money balance)
        {
        return new Transaction(new Id(accountId, transactionNumber), type,
                               new Date(), description, amount, balance);
        }


    // ---- properties ------------------------------------------------------

    public Id getId()
        {
        return m_id;
        }

    public TransactionType getType()
        {
        return m_type;
        }

    public Date getTime()
        {
        return m_time;
        }

    public String getDescription()
        {
        return m_description;
        }

    public Money getAmount()
        {
        return m_amount;
        }

    public Money getBalance()
        {
        return m_balance;
        }


    // ---- Object methods --------------------------------------------------

    public String toString()
        {
        return "Transaction{" +
               "id=" + m_id +
               ", type=" + m_type +
               ", time=" + m_time +
               ", description='" + m_description + '\'' +
               ", amount=" + m_amount +
               ", balance=" + m_balance +
               '}';
        }


    // ---- inner class: Id -------------------------------------------------

    public static class Id
            implements PortableObject, KeyAssociation
        {
        // ---- data members --------------------------------------------

        public Id()
            {
            }

        public Id(long accountId, long transactionNumber)
            {
            m_accountId         = accountId;
            m_transactionNumber = transactionNumber;
            }

        @JsonIgnore
        public Object getAssociatedKey()
            {
            return m_accountId;
            }

        public long getAccountId()
            {
            return m_accountId;
            }

        public long getTransactionNumber()
            {
            return m_transactionNumber;
            }

        public void readExternal(PofReader reader)
                throws IOException
            {
            m_accountId         = reader.readLong(0);
            m_transactionNumber = reader.readLong(1);
            }

        public void writeExternal(PofWriter writer)
                throws IOException
            {
            writer.writeLong(0, m_accountId);
            writer.writeLong(1, m_transactionNumber);
            }

        @Override
        public boolean equals(Object o)
            {
            if (this == o)
                {
                return true;
                }
            if (o == null || getClass() != o.getClass())
                {
                return false;
                }

            Id id = (Id) o;

            return m_accountId == id.m_accountId
                   && m_transactionNumber == id.m_transactionNumber;
            }

        @Override
        public int hashCode()
            {
            int result = (int) (m_accountId ^ (m_accountId >>> 32));
            result = 31 * result + (int) (m_transactionNumber ^ (
                    m_transactionNumber >>> 32));
            return result;
            }

        @Override
        public String toString()
            {
            return "Id(" +
                   "accountId=" + m_accountId +
                   ", transactionNumber=" + m_transactionNumber +
                   ')';
            }

        // ---- data members --------------------------------------------

        private long m_accountId;

        private long m_transactionNumber;
        }


    // ---- inner class: Serializer -----------------------------------------

    public static class Serializer
            extends AbstractPofSerializer<Transaction>
        {
        protected void serializeAttributes(Transaction transaction, PofWriter pofWriter)
                throws IOException
            {
            pofWriter.writeObject(0, transaction.m_id);
            pofWriter.writeObject(1, transaction.m_type);
            pofWriter.writeDate  (2, transaction.m_time);
            pofWriter.writeString(3, transaction.m_description);
            pofWriter.writeObject(4, transaction.m_amount);
            pofWriter.writeObject(5, transaction.m_balance);
            }

        protected Transaction createInstance(PofReader pofReader)
                throws IOException
            {
            return new Transaction(
                    (Id) pofReader.readObject(0),
                    (TransactionType) pofReader.readObject(1),
                    pofReader.readDate(2),
                    pofReader.readString(3),
                    (Money) pofReader.readObject(4),
                    (Money) pofReader.readObject(5));
            }
        }


    // ---- data members ----------------------------------------------------

    private final Id m_id;
    private final TransactionType m_type;
    private final Date m_time;
    private final String m_description;
    private final Money m_amount;
    private final Money m_balance;
    }
